t/mojo/user_agent.t

```diff
@@ -92,432 +90,492 @@ get '/redirect_close' => sub {
+  my $result;
+  $ua->delete_p('/method')->then(sub { $result = shift->res->body })->wait;
+  is $result, 'DELETE', 'right result';
+  $ua->get_p('/method')->then(sub { $result = shift->res->body })->wait;
+  is $result, 'GET', 'right result';
+  $ua->head_p('/method')->then(sub { $result = shift->res->body })->wait;
+  is $result, '', 'no result';
+  $ua->options_p('/method')->then(sub { $result = shift->res->body })->wait;
+  is $result, 'OPTIONS', 'right result';
+  $ua->patch_p('/method')->then(sub { $result = shift->res->body })->wait;
+  is $result, 'PATCH', 'right result';
+  $ua->post_p('/method')->then(sub { $result = shift->res->body })->wait;
+  is $result, 'POST', 'right result';
+  $ua->put_p('/method')->then(sub { $result = shift->res->body })->wait;
+  is $result, 'PUT', 'right result';
+  $ua->start_p($ua->build_tx(TEST => '/method'))->then(sub { $result = shift->res->body })->wait;
+  is $result, 'TEST', 'right result';
+};
+
+subtest 'No timeout' => sub {
+  my $ua = Mojo::UserAgent->new(inactivity_timeout => 0);
+  my $tx = $ua->get('/');
+  ok $tx->keep_alive, 'keep connection alive';
+  is $tx->res->code, 200,      'right status';
+  is $tx->res->body, 'works!', 'right content';
+  $tx = $ua->get('/');
+  ok $tx->kept_alive, 'kept connection alive';
+  ok $tx->keep_alive, 'keep connection alive';
+  is $tx->res->code, 200,      'right status';
+  is $tx->res->body, 'works!', 'right content';
+};
+
+subtest 'SOCKS proxy request without SOCKS support' => sub {
+  my $ua = Mojo::UserAgent->new;
+  my $tx = $ua->build_tx(GET => '/');
+  $tx->req->proxy($ua->server->url->scheme('socks'));
+  $tx = $ua->start($tx);
+  like $tx->error->{message}, qr/IO::Socket::Socks/, 'right error';
+  ok !Mojo::IOLoop::Client->can_socks, 'no SOCKS5 support';
+};
+
+subtest 'HTTPS request without TLS support' => sub {
+  my $ua = Mojo::UserAgent->new;
+  my $tx = $ua->get($ua->server->url->scheme('https'));
+  like $tx->error->{message}, qr/IO::Socket::SSL/, 'right error';
+  ok !Mojo::IOLoop::TLS->can_tls, 'no TLS support';
+};
+
+subtest 'Promises (rejected)' => sub {
+  my $ua = Mojo::UserAgent->new;
+  my $error;
+  $ua->get_p($ua->server->url->scheme('https'))->catch(sub { $error = shift })->wait;
+  like $error, qr/IO::Socket::SSL/, 'right error';
+};
+
+subtest 'No non-blocking name resolution' => sub {
+  ok !Mojo::IOLoop::Client->can_nnr, 'no non-blocking name resolution support';
+};
+
+subtest 'Blocking' => sub {
+  my $ua = Mojo::UserAgent->new;
+  my $tx = $ua->get('/');
+  ok !$tx->error,      'no error';
+  ok !$tx->kept_alive, 'kept connection not alive';
+  ok $tx->keep_alive, 'keep connection alive';
+  is $tx->res->version, '1.1', 'right version';
+  is $tx->res->code,    200,   'right status';
+  ok !$tx->res->headers->connection, 'no "Connection" value';
+  is $tx->res->max_message_size, 2147483648, 'right value';
+  is $tx->res->body,             'works!',   'right content';
+};
+
+subtest 'Again' => sub {
```

Comment:
This is no good: I think the idea here is to make sure that "kept_alive" is true between requests. Not quite sure how to format that though. Maybe a `$ua_keep_alive` object outside of the subtests? Or maybe a "# Again" comment inside the subtest? I guess the latter is better.

Reply:
I'll tell you, I can't tell how much of these tests are actually testing an edge case, or if they were cargo culted from a previous test. So I have no idea what is affecting code coverage here or not.
All of these tests were subtly using the previous state of unrelated tests, which would lead me to believe that they were more boilerplate than real test. But I don't know. I could change it to gratuitous second requests in each of the tests that had to be changed, if that's necessary. But I don't know if it's just noise, as above.

Reply:
Keep-alive tests tend to be significant.

Reply:
so should I add an extra call wherever there was a keep-alive test? My original thought was that kept-alive mirrors keep-alive, so I can just check one of the two.